### PR TITLE
2 chain

### LINKED ISSUE #14: Add highQC in timeout messages
We need to add the highQC in timeout messages to enable bounded time catchup for slow nodes

### PR DESCRIPTION
**Pull request by asonnino**
- Implement 2-chain HotStuff

Closes #14

### CODE REVIEW

File: consensus/src/core.rs [hunk 1]
```diff
@@ -134,16 +135,21 @@ impl<Mempool: 'static + NodeMempool> Core<Mempool> {
     }
 
     async fn make_vote(&mut self, block: &Block) -> Option<Vote> {
+        // Update the preferred round.
+        self.update_preferred_round(block.qc.round);
```
Comment (LefKok) [inline consensus/src/core.rs:139]: Preferred round is not really necessary in a 2-chain, we just have a locked round (high-qc) and a committed round

File: consensus/src/core.rs [hunk 2]
```diff
@@ -174,52 +186,74 @@ impl<Mempool: 'static + NodeMempool> Core<Mempool> {
...
+        }
+
+        // Ensure the timeout is well formed.
+        timeout.verify(&self.committee)?;
+
+        // Process the QC embedded in the timeout.
+        self.process_qc(&timeout.high_qc).await;
+
+        // Add the new vote to our aggregator and see if we have a quorum.
+        if let Some(tc) = self.aggregator.add_timeout(timeout.clone())? {
```
Comment (LefKok) [inline consensus/src/core.rs:220]: If you get a TC you should broadcast it to ensure liveness

File: consensus/src/messages.rs [hunk 3]
```diff
@@ -204,3 +216,109 @@ impl PartialEq for QC {
         self.hash == other.hash && self.round == other.round
     }
 }
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Timeout {
+    pub high_qc: QC,
+    pub round: RoundNumber,
```
Comment (LefKok) [inline consensus/src/messages.rs:223]: For efficiency the timeout signature should be over the round the node is timing out and the round inside the high_qc. This way the leader can forward all O(n) signatures and a single high_qc inside the TC.

**Review by LefKok — APPROVED** [review]

File: consensus/src/core.rs (hunk 2 above)
Comment (asonnino) [reply consensus/src/core.rs:220]: Note: This is also necessary in 3-chain HS!

File: consensus/src/messages.rs (hunk 3 above)
Comment (asonnino) [reply consensus/src/messages.rs:223]: Yes, this is exactly what we do: https://github.com/asonnino/hotstuff/blob/35c6a9985316c1898b53f0d0c37c56a823f30d77/consensus/src/messages.rs#L266

**Pull request merged**